stdlib/seq.mli

```diff
@@ -13,97 +13,807 @@
+    that satisfy [p x].
+
+    In other words, [filter p xs] is the sequence [xs],
+    deprived of the elements [x] such that [p x] is false. *)
 
 val filter_map : ('a -> 'b option) -> 'a t -> 'b t
-(** Apply the function to every element; if [f x = None] then [x] is dropped;
-    if [f x = Some y] then [y] is returned.
-    This transformation is lazy, it only applies when the result is
-    traversed. *)
+(** [filter_map f xs] is the sequence of the elements [y] such that
+    [f x = Some y], where [x] ranges over [xs].
+
+    [filter_map f xs] is equivalent to
+    [map Option.get (filter Option.is_some (map f xs))]. *)
+
+val scan : ('b -> 'a -> 'b) -> 'b -> 'a t -> 'b t
+(** If [xs] is a sequence [[x0; x1; x2; ...]], then
+    [scan f a0 xs] is a sequence of accumulators
+    [[a0; a1; a2; ...]]
+    where [a1] is [f a0 x0], [a2] is [f a1 x1], and so on.
+
+    Thus, [scan f a0 xs] is conceptually related to
+    [fold_left f a0 xs]. However, instead of performing an
+    eager iteration and immediately returning the final accumulator,
+    it returns a sequence of accumulators.
+
+    For instance, [scan (+) 0] transforms a sequence of integers
+    into the sequence of its partial sums.
+
+    If [xs] has length [n]
+    then [scan f a0 xs] has length [n+1].
+
+    @since 4.14 *)
+
+val take : int -> 'a t -> 'a t
+(** [take n xs] is the sequence of the first [n] elements of [xs].
+
+    If [xs] has fewer than [n] elements,
+    then [take n xs] is equivalent to [xs].
+
+    [n] must be nonnegative.
+
+    @raise Invalid_argument if [n] is negative.
+    @since 4.14 *)
+
+val drop : int -> 'a t -> 'a t
+(** [drop n xs] is the sequence [xs], deprived of its first [n] elements.
+
+    If [xs] has fewer than [n] elements,
+    then [drop n xs] is empty.
+
+    [n] must be nonnegative.
+
+    [drop] is lazy: the first [n+1] elements of the sequence [xs]
+    are demanded only when the first element of [drop n xs] is
+    demanded. For this reason, [drop 1 xs] is {i not} equivalent
+    to [tail xs], which queries [xs] immediately.
+
+    @raise Invalid_argument if [n] is negative.
+    @since 4.14 *)
+
+val take_while : ('a -> bool) -> 'a t -> 'a t
+(** [take_while p xs] is the longest prefix of the sequence [xs]
+    where every element [x] satisfies [p x].
+
+    @since 4.14 *)
+
+val drop_while : ('a -> bool) -> 'a t -> 'a t
+(** [drop_while p xs] is the sequence [xs], deprived of the prefix
+    [take_while p xs].
+
+    @since 4.14 *)
+
+val uniq : ('a -> 'a -> bool) -> 'a t -> 'a t
+(** Provided the function [eq] defines an equality on elements,
+    [uniq eq xs] is the sequence [xs],
+    deprived of its adjacent duplicate elements.
+
+    Thus, [uniq eq xs] has no adjacent duplicate elements.
+    Furthermore, if [xs] is sorted according to an ordering
+    that is compatible with [eq],
+    then [uniq eq xs] has no duplicate elements at all.
+
+    [uniq eq xs] is equivalent to [map head (group eq xs)].
+
+    @since 4.14 *)
```

Comment:
I don't like the fact that, to me, the name suggests a *global* uniqueness property for each element. Do you have other suggestions? Maybe `group_heads`, `uniq_in_groups`?

(Also: the documentation does not specify which of the equivalence element is taken, until the "equivalent expression" at the end. The fact that the first element of the group is taken should be written in text as well.)

Reply:
Jane Street's `base` opts for [`remove_consecutive_duplicates`](https://ocaml.janestreet.com/ocaml-core/latest/doc/base/Base/Sequence/index.html#val-remove_consecutive_duplicates), which I find admirable in its straightforwardness :-)

Reply:
It seems this function (and the one following it) could be useful with `eq` functions that are not equalities. For instance, `uniq (>=) xs` could produce the sequence of elements of `xs` that are higher than all previous elements, depending on how it is implemented; however, this would require this function to be a bit more specified (argument order, comparaison to consecutive vs. first-in-a-run elements).